source/common/conn_pool/conn_pool_base.cc

```diff
@@ -600,6 +609,27 @@ void ActiveClient::onConnectTimeout() {
   close();
 }
 
+void ActiveClient::onLifetimeTimeout() {
+  // The lifetime timer should only have started after we left the CONNECTING state.
+  ASSERT(state_ != ActiveClient::State::CONNECTING);
```

Comment:
I looked at this a little harder. If this were to happen in real life, it would be similar to reaching the connect timeout, and the connection would `close()`. However, I think there might be trouble since we're moving the ActiveClient to the busy_clients_ list (not the connecting_clients_ list) and in theory the connect timer could still fire and try to `close()` again. Not entirely sure if either those are problems in practice but it definitely doesn't feel right.

So I think the best two options are to upgrade this to a fatal RELEASE_ASSERT or to keep it as an ASSERT(presumably only fatal in dev builds and tests) and guarding against the bad state by doing nothing if the state is, in fact, CONNECTING. I'm leaning towards the latter because there's no real harm if we hit this state, outside the fact that it means the original feature isn't working correctly :)

Reply:
I'm OK with the latter; can we hit that in a testcase somehow?

Reply:
According to STYLE.md, this case is probably best covered as an ENVOY_BUG and we can test it using EXPECT_ENVOY_BUG.